### PR TITLE
fix: add deadlines for all synchronous REST callers

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
fix: add deadlines for all synchronous REST callers

## Motivation and Context
add deadlines that can be dynamically changed via
the drive max timeout values.

Bonus: optimize "file not found" case and hung drives/network -
circuit, break the check, and return right away instead of waiting.

## How to test this PR?
It allows for the quick detection of hung nodes for all blocking operations.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-object.go [hunk 1]
```diff
@@ -899,6 +899,20 @@ func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object s
...
+			ok := countErrs(errs, errFileNotFound) >= minDisks || countErrs(errs, errFileVersionNotFound) >= minDisks
+			rw.Unlock()
+			if ok {
+				err = errFileNotFound
+				if opts.VersionID != "" {
+					err = errFileVersionNotFound
+				}
+				break
+			}
+		}
```
Comment (harshavardhana) [inline cmd/erasure-object.go:914]: The idea is that the return file is not found faster; this is better for `Head` like calls.

File: internal/config/drive/drive.go [hunk 2]
```diff
@@ -25,6 +25,10 @@ import (
 	"github.com/minio/pkg/v2/env"
 )
 
+const (
+	EnvMaxDriveTimeout = "MINIO_DRIVE_MAX_TIMEOUT"
```
Comment (klauspost) [inline internal/config/drive/drive.go]: What does the "MAX" refer to? Maybe replace with `OP` or `OPERATTION`  now you have to change it for the linter?

**Review by klauspost — COMMENTED** [review]
LGTM

File: cmd/storage-rest-client.go [hunk 3]
```diff
@@ -391,6 +391,9 @@ func (client *storageRESTClient) CreateFile(ctx context.Context, origvolume, vol
 }
 
 func (client *storageRESTClient) WriteMetadata(ctx context.Context, origvolume, volume, path string, fi FileInfo) error {
+	ctx, cancel := context.WithTimeout(ctx, globalDriveConfig.GetMaxTimeout())
```
Comment (vadmeste) [inline cmd/storage-rest-client.go:394]: isn't ping/pong in the grid code enough ? probably we should fix something in the grid code instead

Comment (klauspost) [reply cmd/storage-rest-client.go:394]: @vadmeste Timeouts are strict timeouts and are forwarded to the remote. So both the caller and the remote will hit this timeout.

Ping will keep the request alive forever even with no activity - and are only available on streams.
Same with context cancelation - only for streams - Streams will propagate cancelation.

One-shot requests are assumed to be short lived - and have a default deadline of 1 minute, unless something else is set on the context. Cancelling the context will cancel the local, but the remote will have to hit the timeout.

Comment (harshavardhana) [reply cmd/storage-rest-client.go:394]: We cannot rely on pings. Active callers must timeout.

File: internal/config/drive/drive.go (hunk 2 above)
Comment (harshavardhana) [reply internal/config/drive/drive.go]: I will send a separate PR for it

**Review by krisis — APPROVED** [review]
LGTM

**Pull request merged**